File: src/dji_motor.py
import struct
from motor_base import MotorBase

class DJIMotor(MotorBase):
    def __init__(self, bus_id, motor_id, model):
        """
        DJI 电机驱动类 (支持 GM6020 / GM3508)
        
        :param bus_id: 挂载的 CAN 总线 ID (0, 1...)
        :param motor_id: 电机物理 ID (拨码开关设定值)
        :param model: 电机型号字符串 "GM6020" 或 "GM3508"
        """
        super().__init__(bus_id, motor_id)
        self.model = model
        self.target_output = 0 # 缓存 PID 计算后的控制量 (电压或电流)

        # === ID 映射与参数配置 ===
        
        if model == "GM6020":
            # GM6020 (电压控制)
            # 反馈 ID: ID 1 -> 0x205 ...
            self.feedback_id = 0x205 + (motor_id - 1)
            
            # 控制 ID: 
            # ID 1-4 -> 0x1FF
            # ID 5-8 -> 0x2FF
            self.control_id = 0x1FF if motor_id <= 4 else 0x2FF
            
            # 限幅 (电压)
            self.max_out = 30000

        elif model == "GM3508":
            # GM3508 (电流控制, C620电调)
            # 反馈 ID: ID 1 -> 0x201 ...
            self.feedback_id = 0x201 + (motor_id - 1)
            
            # 控制 ID:
            # ID 1-4 -> 0x200
            # ID 5-8 -> 0x1FF  (注意：这里会与 GM6020 ID 1-4 冲突，需物理避开)
            self.control_id = 0x200 if motor_id <= 4 else 0x1FF
            
            # 限幅 (电流)
            self.max_out = 16384
        
        else:
            raise ValueError(f"[DJIMotor] 未知型号: {model}")

    def update_feedback(self, can_id, data):
        """
        解析 DJI 标准反馈帧 (8字节)
        格式 (Big Endian): Angle(2) RPM(2) Current(2) Temp(1) Null(1)
        """
        # 严格校验 CAN ID
        if can_id != self.feedback_id: 
            return

        self.online = True # 标记电机在线
        
        # 解包
        (raw_angle, raw_rpm, raw_cur, temp) = struct.unpack('>hhhB', data[:7])
        
        # 更新状态属性 (MotorBase 属性)
        # 将 0-8191 映射为 0-360 度
        self.angle = (raw_angle / 8191.0) * 360.0
        self.speed = raw_rpm
        self.torque = raw_cur
        self.temp = temp

    def set_output(self, value):
        """
        设置目标输出值 (由 PID 计算得出)
        :param value: GM6020为电压值, GM3508为电流值
        """
        # 通用限幅逻辑
        if value > self.max_out: value = self.max_out
        if value < -self.max_out: value = -self.max_out
        
        self.target_output = int(value)

    def get_control_bytes(self):
        """
        获取 2 字节控制数据 (Big Endian)
        用于 Main Loop 中的拼包逻辑 (0x1FF / 0x200)
        """
        val = self.target_output
        return [(val >> 8) & 0xFF, val & 0xFF]

    def get_control_frame(self):
        """
        MotorBase 接口实现
        注: DJI 电机通常使用共享帧 (0x1FF/0x200) 统一发送，
        不建议单电机独立发送，因此这里返回 None。
        发送逻辑由 Main Loop 里的拼包器接管。
        """
        return None, None

File: src/motor_control_ros2/src/dji_motor.cpp
#include "motor_control_ros2/dji_motor.hpp"
#include <cstring>
#include <iostream>
#include <iomanip>
#include <chrono>

namespace motor_control {

DJIMotor::DJIMotor(const std::string& joint_name, MotorType type, 
                   uint8_t motor_id, uint8_t bus_id)
  : MotorBase(joint_name, type, 
              (type == MotorType::DJI_GM3508) ? 19.0f : 1.0f,  // 减速比
              (type == MotorType::DJI_GM6020))  // GM6020 编码器在输出轴，GM3508 编码器在电机轴
  , motor_id_(motor_id)
  , bus_id_(bus_id)
  , target_output_(0)
  , raw_angle_(0)
  , raw_rpm_(0)
  , raw_current_(0)
  , raw_temp_(0)
  , last_raw_angle_(0)
  , encoder_rounds_(0)
  , first_feedback_(true)
  , position_target_(0.0)
  , velocity_target_(0.0)
{
  if (type == MotorType::DJI_GM6020) {
    // GM6020 配置
    feedback_id_ = 0x205 + (motor_id - 1);
    control_id_ = (motor_id <= 4) ? 0x1FF : 0x2FF;
    max_output_ = 30000;  // 电压限幅
  } else if (type == MotorType::DJI_GM3508) {
    // GM3508 配置
    feedback_id_ = 0x201 + (motor_id - 1);
    control_id_ = (motor_id <= 4) ? 0x200 : 0x1FF;
    max_output_ = 16384;  // 电流限幅
  }
  
  // 默认为直接输出模式
  cascade_controller_.setMode(ControlMode::DIRECT);
}

void DJIMotor::updateFeedback(const std::string& interface_name, uint32_t can_id, const uint8_t* data, size_t len) {
  // 检查接口名称匹配（多路 CAN 支持）
  if (!interface_name_.empty() && interface_name != interface_name_) {
    return;  // 不是该电机所属的接口
  }
  
  if (can_id != feedback_id_ || len < 8) {
    return;
  }
  
  // 更新心跳时间（会自动设置 online_ = true）
  updateLastFeedbackTime(std::chrono::steady_clock::now().time_since_epoch().count());
  
  // 解析 DJI 反馈帧 (Big Endian)
  raw_angle_ = (static_cast<uint16_t>(data[0]) << 8) | data[1];
  raw_rpm_ = (static_cast<int16_t>(data[2]) << 8) | data[3];
  raw_current_ = (static_cast<int16_t>(data[4]) << 8) | data[5];
  raw_temp_ = data[6];
  
  // 累积编码器逻辑（用于 GM3508，编码器在电机轴）
  if (!encoder_on_output_) {
    // GM3508: 编码器在电机轴，需要累积圈数
    if (!first_feedback_) {
      // 检测过零
      int32_t delta = static_cast<int32_t>(raw_angle_) - static_cast<int32_t>(last_raw_angle_);
      
      // 过零检测：如果变化超过半圈（4096），说明发生了过零
      if (delta > 4096) {
        // 从 8191 跳到 0（逆时针过零）
        encoder_rounds_--;
      } else if (delta < -4096) {
        // 从 0 跳到 8191（顺时针过零）
        encoder_rounds_++;
      }
    }
    
    last_raw_angle_ = raw_angle_;
    first_feedback_ = false;
    
    // 计算累积位置（电机轴）= 累积圈数 × 2π + 当前圈内位置
    double rounds_position = encoder_rounds_ * 2.0 * M_PI;
    double current_round_position = (raw_angle_ / 8191.0) * 2.0 * M_PI;
    position_ = rounds_position + current_round_position;
    
  } else {
    // GM6020: 编码器在输出轴，直接转换
    position_ = (raw_angle_ / 8191.0) * 2.0 * M_PI;
  }
  
  // 速度转换（两种电机都一样，RPM 是电机轴速度）
  velocity_ = (raw_rpm_ / 60.0) * 2.0 * M_PI;
  torque_ = raw_current_;
  temperature_ = static_cast<float>(raw_temp_);
  
  // 保持反馈解析热路径简洁，避免在此输出周期性 stdout 日志
}

void DJIMotor::getControlFrame(uint32_t& can_id, uint8_t* data, size_t& len) {
  // DJI 电机使用共享帧，不单独发送
  // 返回 len = 0 表示需要拼包
  (void)data; // 消除未使用参数警告
  can_id = control_id_;
  len = 0;
}

void DJIMotor::disable() {
  setOutput(0);
}

void DJIMotor::setOutput(int16_t value) {
  // 限幅
  if (value > max_output_) value = max_output_;
  if (value < -max_output_) value = -max_output_;
  target_output_ = value;
}

void DJIMotor::getControlBytes(uint8_t bytes[2]) const {
  // Big Endian
  bytes[0] = (target_output_ >> 8) & 0xFF;
  bytes[1] = target_output_ & 0xFF;
}

// ========== 串级控制实现 ==========

void DJIMotor::setControlMode(ControlMode mode) {
  cascade_controller_.setMode(mode);
}

ControlMode DJIMotor::getControlMode() const {
  return cascade_controller_.getMode();
}

void DJIMotor::setPositionTarget(double position) {
  position_target_ = position;
}

void DJIMotor::setVelocityTarget(double velocity) {
  velocity_target_ = velocity;
}

void DJIMotor::setPositionPID(const PIDParams& params) {
  cascade_controller_.setPositionPID(params);
}

void DJIMotor::setVelocityPID(const PIDParams& params) {
  cascade_controller_.setVelocityPID(params);
}

void DJIMotor::updateController() {
  // 使用度和 RPM 作为单位，与 Python 实现一致
  double output = cascade_controller_.update(
    position_target_,                    // 位置目标 (度)
    velocity_target_,                    // 速度目标 (RPM)
    static_cast<double>(target_output_), // 直接输出
    getAngleDegrees(),                   // 位置反馈 (度, 0-360)
    static_cast<double>(raw_rpm_)        // 速度反馈 (RPM)
  );
  
  // 将输出转换为 int16 并设置
  setOutput(static_cast<int16_t>(output));
}

} // namespace motor_control
